src/qoi/src/integrated_function.C

```diff
@@ -0,0 +1,140 @@
+#include "grins/rayfire_mesh.h"
+
+// libMesh
+#include "libmesh/getpot.h"
+#include "libmesh/fem_system.h"
+#include "libmesh/quadrature_gauss.h"
+#include "libmesh/elem.h"
+#include "libmesh/fe.h"
+#include "libmesh/fe_type.h"
+#include "libmesh/function_base.h"
+#include "libmesh/fem_function_base.h"
+
+namespace GRINS
+{
+  template<typename func>
+  IntegratedFunction<func>::IntegratedFunction(unsigned int p_level, SharedPtr<func> f, SharedPtr<RayfireMesh> rayfire, const std::string& qoi_name) :
+    QoIBase(qoi_name),
+    _p_level(p_level),
+    _f(f),
+    _rayfire(rayfire)
+  {
+    // use Gauss Quadrature
+    _qbase = new libMesh::QGauss(1,libMesh::Order(_p_level));
+  }
+
+  template<typename func>
+  QoIBase* IntegratedFunction<func>::clone() const
+  {
+    return new IntegratedFunction<func>( *this );
+  }
+
+  template<typename func>
+  void IntegratedFunction<func>::init
+    (const GetPot& /*input*/,
+     const MultiphysicsSystem& system,
+     unsigned int /*qoi_num*/ )
+  {
+    _rayfire->init(system.get_mesh());
+  }
+
+  template<typename func>
+  void IntegratedFunction<func>::init_context( AssemblyContext& /*context*/ )
+  {}
+
+  template<typename func>
+  void IntegratedFunction<func>::element_qoi( AssemblyContext& context,
+                                       const unsigned int qoi_index )
+  {
+    const libMesh::Elem& original_elem = context.get_elem();
+    const libMesh::Elem* rayfire_elem = _rayfire->map_to_rayfire_elem(original_elem.id());
+
+    // rayfire_elem will be NULL if the main_elem
+    // is not in the rayfire
+    if (rayfire_elem)
+      {
+        // init the quadrature base on the rayfire elem
+        _qbase->init(rayfire_elem->type(),_p_level);
+
+        // need the QP coordinates and JxW
+        libMesh::UniquePtr< libMesh::FEBase > fe = libMesh::FEBase::build(rayfire_elem->dim(), libMesh::FEType(libMesh::FIRST, libMesh::LAGRANGE) );
+
+        fe->attach_quadrature_rule( _qbase.get() );
+        fe->get_xyz();
+        fe->get_JxW();
+
+        fe->reinit(rayfire_elem);
+
+        const std::vector<libMesh::Real>& JxW = fe->get_JxW();
+        const std::vector<libMesh::Point>& xyz = fe->get_xyz();
+
+        const unsigned int n_qpoints = fe->n_quadrature_points();
+
+        libMesh::Number& qoi = context.get_qois()[qoi_index];
+
+        for (unsigned int qp = 0; qp != n_qpoints; ++qp)
+            qoi += this->qoi_value((*_f),context,xyz[qp])*JxW[qp];
+
+      }
+  }
+
+  template<typename func>
+  void IntegratedFunction<func>::element_qoi_derivative( AssemblyContext& /*context*/,
+                                       const unsigned int /*qoi_index*/ )
+  {
+    //TODO
+    libmesh_not_implemented();
+  }
+
+
+  // private
+
+  // speciaizations of the qoi_value() function
+  template<>
+  libMesh::Real IntegratedFunction<libMesh::FEMFunctionBase<libMesh::Real>>::qoi_value(libMesh::FEMFunctionBase<libMesh::Real>& f,AssemblyContext& context,const libMesh::Point& xyz)
```

Comment:
We're not officially C++11 only yet, so you need a space between the two `>>`, i.e. `IntegratedFunction<libMesh::FEMFunctionBase<libMesh::Real>>` should be `IntegratedFunction<libMesh::FEMFunctionBase<libMesh::Real> >`